<br>
<table style="font-family: Helvetica,Arial,sans-serif; width: 99%; text-align: left; margin-left: auto; margin-right: auto;" border="0" cellpadding="10" cellspacing="2">
<tbody>
<tr style="font-weight: bold;" class="head">
<td style="width: 33%; background-color: rgb(255, 255, 204);" colspan="4" rowspan="1">Gene Set Enrichment Analysis</td>
</tr>
<tr class="line1">
<td colspan="2" rowspan="1" style="background-color: rgb(205, 213, 255); width: 33%; text-align: left;">
<div style="text-align: justify;">Pathway analysis
methods focus on the behavior of a gene set rather
than individual genes by evaluating gene expression profiles of
a-priori defined biological pathways in association with a phenotype of
interest. We will use Gene Set Enrichment Analysis (GSEA ) a
non-parametric annotation-driven statistical analysis method. The basic
idea of GSEA is to compare two
phenotypes to determine whether a known biological pathway or sets of
individual genes are significantly enriched in a list of genes ranked
according to criteria such as the expression difference between the two
phenotypes of interest.&nbsp;<br>
<br>
We
will use and test systematically three Categories of gene sets from the
Molecular signature Database (MsigDB,
<a href="http://www.broad.mit.edu/gsea/msigdb">http://www.broad.mit.edu/gsea/msigdb</a>) version 3.0 (Oct. 2010)
composed of:<br>
1)&nbsp;&nbsp;&nbsp; 880 gene
sets (referenced as C2) collected from various sources such as online
pathway databases, publications in PubMed, and knowledge of domain
experts<br>
2)&nbsp;&nbsp; 615 gene sets (referenced as C3) that
contain genes that
share a cis-regulatory motif that is conserved across the human, mouse,
rat, and dog genomes and &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; represent known or predicted
regulatory
elements
in promoters and 3'-UTRs matching transcription factor targets. <br>
3)&nbsp;&nbsp; 825 gene sets of Biological Processes (referenced as C5) coming from the Gene Ontology Project.<br>
<br>
Merged
to C2 are 28 gene sets collected from transcriptional analyses of the
blood in immunological context, and are named M x.x (eg. M 1.3). Their
descriptions are not available through MsigDB, but can be found in
Chaussabel et al. (2008) :<br><div style="margin-left: 40px;"><a href="blood_modules.pdf">A Modular Analysis Framework for
Blood Genomics Studies</a>: Application to Systemic Lupus
Erythematosus<br>
Immunity, Volume 29, Issue 1, Pages 150-164<br></div><br>GSEA separate significantly regulated pathways between <span style="font-weight: bold;">upregulated</span> and <span style="font-weight: bold;">downregulated pathways</span>. For each contrast and each database, you can browse the regulated pathways sorted by decreasing significance. We also provide a <span style="font-weight: bold;">Summary Figure</span>, which represent the expression levels accross samples relevant to the contrast for the top 5 most significant genes from the top 5 up and the top 5 downregulated pathways. The pathway membership for all genes is represented in the left panel of the figure. </div>
</td>
</tr>
<tr class="line2">
<td colspan="2"><a href="howto.html">Tutorial</a> for browsing the GSEA results <span style="color: rgb(142, 142, 142);">(under
construction)</span></td>
</tr>
</tbody>
</table>
<br>
